src/compiler/checker.ts

```diff
@@ -20233,16 +20238,35 @@ namespace ts {
                         mapper = createTypeMapper(typeParameters, typeArguments);
                     }
                     const typeArgument = typeArguments[i];
+                    const instantiatedConstraint = instantiateType(constraint, mapper);
+                    if (constraints) constraints[i] = instantiatedConstraint;
                     result = result && checkTypeAssignableTo(
                         typeArgument,
-                        instantiateType(constraint, mapper),
+                        instantiatedConstraint,
                         typeArgumentNodes[i],
                         Diagnostics.Type_0_does_not_satisfy_the_constraint_1);
                 }
             }
             return result;
         }
 
+        function getTypeArgumentConstraint(node: TypeNode): Type | undefined {

```

Comment:
I think getTypeArgumentConstraint should not call `checkTypeArgumentConstraints`, but just call `getConstraintOfTypeParameter` directly. Then it simplifies to:

```ts
const typeReferenceNode = ...
const typeParameters = ...
// (other early exit code)
return getConstraintOfTypeParameter(typeParameters[typeReferenceNode.typeArguments.indexOf(node)]);
```

The instantiation code shouldn't be needed because typeArguments will always be defined. I'm not sure what having `mapper === undefined` does anyway.

Reply:
Having `mapper === undefined` causes problems if you have `Pick<Foo, "">` -- you want to get `keyof Foo` and not `keyof T`.

Reply:
Where does `checkTypeArgumentContraints` set `mapper`s value for the case that `typeArguments` is defined?

Reply:
Oh, never mind, `typeArguments` !== `typeArgumentNodes` and is lazily initialized. Well, I would still clone those two lines into `getTypeArgumentConstraint`.

Reply:
Maybe there's some way to share the loop body between the two, although I don't see anything good.

Reply:
Looks like the loop isn't necessary, we were just computing the mapper lazily but it doesn't really matter when we do.